### PR TITLE
Support dollar sign as query parameter value in simple queries

### LINKED ISSUE #312: $1, $2, … in string literals erroneously identified as bind marker
#### Versions

- Driver: io.r2dbc/r2dbc-postgresql
- Database: userspasswords 
- Java: 11
- OS: Windows 10 x64

#### Current Behavior
The Special character "$" are not being accepted for the migration process. 

<details>
<summary>Stack trace</summary>

```

By 'Writing metadata version 3' 1 rows updated
web-api_1     | 2020-08-04 20:54:43.971  INFO 7 --- [tor-tcp-epoll-1] n.n.r2dbc.migrate.core.R2dbcMigrate      : Applying MigrationInfo{version=4, description='insert in table userpasswords', splitByLine=false, transactional=true}
web-api_1     | 2020-08-04 20:54:43.977  INFO 7 --- [tor-tcp-epoll-1] n.n.r2dbc.migrate.core.R2dbcMigrate      : By 'Releasing lock after error' 1 rows updated
web-api_1     | 2020-08-04 20:54:43.980  WARN 7 --- [           main] onfigReactiveWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'animeController': Unsatisfied dependency expressed through field 'service'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'animeService': Unsatisfied dependency expressed through field 'repo'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'animeRepository' defined in academy.devdojo.webflux.repository.AnimeRepository defined in @EnableR2dbcRepositories declared on R2dbcRepositoriesAutoConfigureRegistrar.EnableR2dbcRepositoriesConfiguration: Cannot resolve reference to bean 'r2dbcDatabaseClient' while setting bean property 'databaseClient'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.springframework.boot.autoconfigure.data.r2dbc.R2dbcDataAutoConfiguration': Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'r2dbcMigrate' defined in class path resource [name/nkonev/r2dbc/migrate/autoconfigure/R2dbcMigrateAutoConfiguration.class]: Invocation of init method failed; nested exception is java.lang.IllegalArgumentException: Statement 'INSERT INTO userspasswords (nameuser, username, password, authorities)
web-api_1     | VALUES ('paulo',
web-api_1     |         'paulo',
web-api_1     |         '{bcrypt}$2a$10$lBm1Qy45bR/fNT5i5OUBseNPcONtZs10earjLZ773qq.byhK/yKmS',
web-api_1     |         'ROLE_ADMIN,ROLE_USER');
web-api_1     |
web-api_1     | INSERT INTO userspasswords (nameuser, username, password, authorities)
web-api_1     | VALUES ('demetria',
web-api_1     |         'demetria',
web-api_1     |         'devdojo',
web-api_1     |         'ROLE_USER');' cannot be created. This is often due to the presence of both multiple statements and parameters at the same time.
web-api_1     | 2020-08-04 20:54:43.991  INFO 7 --- [           main] ConditionEvaluationReportLoggingListener :
web-api_1     |
web-api_1     | Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
web-api_1     | 2020-08-04 20:54:43.996 ERROR 7 --- [           main] o.s.boot.SpringApplication               : Application run failed


```

</details>
 
#### Table schema

<details>
<summary>Input Code</summary>

```sql

CREATE TABLE IF NOT EXISTS userspasswords
(
    id SERIAL PRIMARY KEY,
    nameuser VARCHAR(255) NOT NULL,
    username VARCHAR(255) NOT NULL,
    password VARCHAR(255) NOT NULL,
    authorities VARCHAR(255) NOT NULL
);

INSERT INTO userspasswords (nameuser, username, password, authorities)
VALUES ('paulo',
        'paulo',
        'devdojo',
        'ROLE_ADMIN,ROLE_USER');
--         Special character $ area not allowed in the field password
--         '{bcrypt}$2a$10$lBm1Qy45bR/fNT5i5OUBseNPcONtZs10earjLZ773qq.byhK/yKmS',

```

</details>

#### Steps to reproduce

0) GitHub project link: https://github.com/PauloPortfolio/springwebflux-tdd-devdojo 

1) For example: This INSERT statement is completely normal, and it is normal processed for the r2dbc-migrate BECAUSE we have a 'plaintext' password (DEVDOJO):

```
INSERT INTO userspasswords 
          (nameuser, username, password, authorities)
VALUES ('demetria',
        'demetria',
        'devdojo',
        'ROLE_USER');
```

2) HOWEVER, when I change the field 'PASSWORD' from PLAIN-TEXT 'DEVDOJO' 
to ENCRYTED PASSWORD 'DEVDOJO' ({bcrypt}$2a$10$lBm1Qy45bR/fNT5i5OUBseNPcONtZs10earjLZ773qq.byhK/yKmS), the specials characteres are not allowed, specially "$", for instance (THE ERRAR IN THE ABOVE STACK-TRACE HAPPENS):

```
INSERT INTO userspasswords 
          (nameuser, username, password, authorities)
VALUES ('demetria',
        'demetria',
        '{bcrypt}$2a$10$lBm1Qy45bR/fNT5i5OUBseNPcONtZs10earjLZ773qq.byhK/yKmS',
        'ROLE_USER');
```

<details>
<summary>Input Code</summary>

</details>

#### Expected behavior/code

Special characteres chould be allowed, such as ( `'{bcrypt}$2a$10$lBm1Qy45bR/fNT5i5OUBseNPcONtZs10earjLZ773qq.byhK/yKmS'`), because it is acommon practice in the PostGres DB use.

### PR DESCRIPTION
**Pull request by Isabek**
[resolves #312]

Make sure that:

- [x] You have read the [contribution guidelines](https://github.com/pgjdbc/r2dbc-postgresql/blob/main/.github/CONTRIBUTING.adoc).
- [x] You have created a feature request first to discuss your contribution intent. Please reference the feature request ticket number in the pull request.
- [x] You use the code formatters provided [here](https://github.com/pgjdbc/r2dbc-postgresql/blob/master/intellij-style.xml) and have them applied to your changes. Don't submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.

### CODE REVIEW

**Comment by Isabek** [comment]
@mp911de could you review it?

File: src/main/java/io/r2dbc/postgresql/SimpleQueryPostgresqlStatement.java [hunk 1]
```diff
@@ -136,7 +136,18 @@ public String toString() {
     static boolean supports(String sql) {
         Assert.requireNonNull(sql, "sql must not be null");
 
-        return sql.trim().isEmpty() || !sql.contains("$1");
+        boolean singleQuoteOpened = false;
```
[the remark this answers is not in the document]
Comment (Isabek) [reply src/main/java/io/r2dbc/postgresql/SimpleQueryPostgresqlStatement.java:139]: @nkonev thanks for the feedback. What about the following case `$SomeTag$Dianne's horse$SomeTag$`? According to the documentation, this is also a valid case. Should I handle it too?

[the remark this answers is not in the document]
Comment (mp911de) [reply src/main/java/io/r2dbc/postgresql/SimpleQueryPostgresqlStatement.java:139]: Yes, we should support what Postgres allows.

[the remark this answers is not in the document]
Comment (Isabek) [reply src/main/java/io/r2dbc/postgresql/SimpleQueryPostgresqlStatement.java:139]: @mp911de noted, thanks.

[the remark this answers is not in the document]
Comment (nkonev) [reply src/main/java/io/r2dbc/postgresql/SimpleQueryPostgresqlStatement.java:139]: @Isabek thank you for spending your time for this

**Comment by mp911de** [comment]
Is this one ready for merge?

**Comment by Isabek** [comment]
> Is this one ready for merge?

Right now it doesn't support the following format `$SomeTag$Dianne's horse$SomeTag$`.

**Comment by mp911de** [comment]
Any update here?

**Comment by Isabek** [comment]
@mp911de I will take a look in the upcoming two weeks.

**Comment by mp911de** [comment]
Any update here?

**Comment by mp911de** [comment]
Closing due to lack of requested feedback. If you would like us to look at this issue, please provide the requested information and we will re-open the issue.

**Pull request closed without merging**